ql/src/java/org/apache/hadoop/hive/ql/ddl/table/info/show/tables/ShowTablesOperation.java

```diff
@@ -78,20 +74,20 @@ private void showTables() throws HiveException {
   }
 
   private void showTablesExtended() throws HiveException {
-    List<Table> tableObjects = new ArrayList<>();
-    tableObjects.addAll(context.getDb().getTableObjects(desc.getDbName(), null, desc.getTypeFilter()));
-    if (desc.getPattern() != null) {
-      Pattern pattern = Pattern.compile(UDFLike.likePatternToRegExp(desc.getPattern()), Pattern.CASE_INSENSITIVE);
-      tableObjects = tableObjects.stream()
-          .filter(object -> pattern.matcher(object.getTableName()).matches())
-          .collect(Collectors.toList());
+    TreeMap<String, String> tableNameToType = new TreeMap<>();
+    String pattern = MetaStoreUtils.convertSqlPatternToRegExp(desc.getPattern());
+    TableType typeFilter = desc.getTypeFilter();
+    TableType[] tableTypes = typeFilter == null ? TableType.values() : new TableType[]{typeFilter};
+    for (TableType tableType : tableTypes) {
+      String typeString = tableType.toString();
+      List<String> tables = context.getDb().getTablesByType(desc.getDbName(), pattern, tableType);
+      tables.forEach(name -> tableNameToType.put(name, typeString));
```

Comment:
minor: could we inline local var: `tableNameToType.put(name, tableType.toString())` ?